lib/composite_cache_store.rb

```diff
@@ -4,146 +4,164 @@
 require_relative "composite_cache_store/version"
 
 class CompositeCacheStore
-  DEFAULT_LAYER_1_OPTIONS = {
-    expires_in: 5.minutes,
-    size: 16.megabytes
-  }
-
-  DEFAULT_LAYER_2_OPTIONS = {
-    expires_in: 1.day,
-    size: 32.megabytes
-  }
-
-  attr_reader :layers
+  attr_reader :options, :layers
+  attr_accessor :logger
 
   # Returns a new CompositeCacheStore instance
-  def initialize(*layers)
+  def initialize(options = {})
+    options = options.dup || {}
+    layers = options.delete(:layers) || []
     if layers.blank?
-      layers << ActiveSupport::Cache::MemoryStore.new(DEFAULT_LAYER_1_OPTIONS)
-      layers << ActiveSupport::Cache::MemoryStore.new(DEFAULT_LAYER_2_OPTIONS)
+      layers << ActiveSupport::Cache::MemoryStore.new({expires_in: 5.minutes, size: 16.megabytes}.merge(options))
+      layers << ActiveSupport::Cache::MemoryStore.new({expires_in: 1.day, size: 32.megabytes}.merge(options))
     end
 
-    message = "All layers must be instances of ActiveSupport::Cache::Store"
-    layers.each do |layer|
-      raise ArgumentError.new(message) unless layer.is_a?(ActiveSupport::Cache::Store)
-    end
+    stores = layers.select { |layer| layer.is_a? ActiveSupport::Cache::Store }
+    raise ArgumentError.new("All layers must be instances of ActiveSupport::Cache::Store") unless stores.size == layers.size
 
-    layers.freeze
-    @layers = layers
+    @layers = layers.freeze
+    @logger = options[:logger]
+    @options = options
   end
 
   def cleanup(...)
-    layers.each { |store| store.cleanup(...) }
+    layers.each { |layer| layer.cleanup(...) }
   end
 
   def clear(...)
-    layers.each { |store| store.clear(...) }
+    layers.each { |layer| layer.clear(...) }
   end
 
-  def decrement(...)
-    layers.each { |store| store.decrement(...) }
+  def increment(name, amount = 1, options = nil)
+    value = layers.last.increment(name, amount, options)
+    provisional_layers.each { |layer| layer.write(name, value, options) }
+    value
+  end
+
+  def decrement(name, amount = 1, options = nil)
+    value = layers.last.decrement(name, amount, options)
+    provisional_layers.each { |layer| layer.write(name, value, options) }
+    value
```

Comment:
I'm going to dissent here. I think incr/decr are usually used when performance is vital and avoiding race conditions is important, e.g. Rack::Attack counting hits. There's no real way to implement this that keeps those features. `write`ing to intermediate layers, in the 99% case, only slows things down. And getting correct results without race conditions means most apps will almost never get/fetch the data, only use the result returned from incr/decr.

One valid option would be to just do what ActiveSupport::Cache does and [not support them](https://github.com/rails/rails/blob/8015c2c2cf5c8718449677570f372ceb01318a32/activesupport/lib/active_support/cache.rb#L565). Force the caller to invoke Rails.cache.incr (or whatever the outer layer is) since they should just be using it anyway.

Another valid option would be to pass thru to the outer layer as a convenience. If the caller does end up doing a get/fetch, the inner layers can get populated on demand:

```suggestion
  def increment(name, amount = 1, options = nil)
    layers.last.increment(name, amount, options)
  end

  def decrement(name, amount = 1, options = nil)
    layers.last.decrement(name, amount, options)
```

Reply:
> Another valid option would be to pass thru to the outer layer as a convenience. If the caller does end up doing a get/fetch, the inner layers can get populated on demand

The problem with this is that we still need to remove inner layer values which is basically what I'm doing. I suppose the only real options are.

1. Dont' support increment/decrement
2. Delelete inner layer values when these methods are called (let inner layers lazily hydrate if accessed via fetch)
3. Write inner layer values when these methods are called (current implementation)

I sort of like option 1.

Reply:
Given the number of tests this change broke, I decided to keep increment/decrement but switched to deleting the key from provisional layers. Not perfect, but an acceptable compromise for now. We can follow up with a different solution if this proves to be a real performance issue.